packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts

```diff
@@ -82,37 +85,41 @@ export class PnpSignAction implements Action<SignMessageRequest> {
         }
       }
 
-      const key: Key = {
-        version:
-          this.io.getRequestKeyVersion(session.request, session.logger) ??
-          this.config.keystore.keys.phoneNumberPrivacy.latest,
-        name: DefaultKeyName.PHONE_NUMBER_PRIVACY,
-      }
-
       // Compute signature inside transaction so it will rollback on error.
-
+      // Since it's possible that the trx was already rolled back in
+      // checkAndUpdateQuotaStatus, catch and handle any additional errors thrown,
+      // otherwise the signer will hang.
       try {
```

Comment:
Turns out that errors thrown after the rollback do not throw (+ is not caught) by the outer error handling function (since calling rollback explicitly already causes the promise to reject; by default an error that triggers an auto-rollback also throws an error, but calling trx.rollback() (without passing in an error) doesn't). This means we need to send responses in this function itself otherwise the signer doesn't respond anything. An alternative here is having checkAndUpdateQuotaStatus return a boolean indicating whether the outer function _should_ rollback and then rolling back at the end, but that feels just as messy/tricky tbh :-/ Open to other ideas here!

Reply:
Hmm okay I think I see what you mean now. My concern with this current approach is that we resolve the type of ErrorMessage we should return in a way that's difficult to reason about. My vote might be to rely on auto-rollback wherever possible and only use trx.rollback() when we are sending a response. There's definitely some nuance here so happy to jump on a call to discuss further

Reply:
It does seem like using trx.rollback() should probably be avoided given this behavior though

Reply:
I think it could make sense to discuss this sync! Condensed from convo on slack but:

> My vote might be to rely on auto-rollback wherever possible and only use trx.rollback() when we are sending a response.

I think we're currently still trying to do this as much as possible; unless we remove the rollback earlier in the code (allowing fail open but not storing queries if performed queries are incremented), I can't see a way of not needing to manually handle possible later rollbacks during key fetching/signing :-/

Reply:
Updated to now
- propagate DB update errors (when storing requests or incrementing query counts)
  - default to controller's error handling for DB & key signing errors (internally, we should be able to debug from logs, but the combiner will receive UNKNOWN_ERROR (500)'s)
- throw a SIGNATURE_COMPUTATION_ERROR when anything fails in that sig computation try/catch (previously we would check the error message, but I think this is a bit simpler while providing a generally helpful hint about where things went wrong)